flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/ZkLockFactory.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.api;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.apache.iceberg.jdbc.UncheckedSQLException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.zookeeper.KeeperException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Zookeeper backed implementation of the {@link TriggerLockFactory}. */
+public class ZkLockFactory implements TriggerLockFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(ZkLockFactory.class);
+
+  private static final String LOCK_BASE_PATH = "/iceberg/flink/maintenance/locks";
+
+  private final String connectString;
+  private final String lockId;
+  private final int sessionTimeoutMs;
+  private final int connectionTimeoutMs;
+  private final ExponentialBackoffRetryWrapper retryWrapper;
+  private transient CuratorFramework client;
+
+  /**
+   * Create Zookeeper lock factory
+   *
+   * @param connectString Zookeeper connection string
+   * @param lockId Lock ID to distinguish different jobs
```

Comment:
this is for distinguish different tables. So no 2 jobs could run table maintenance on the same table